feature: backward compatibility validation for Conjure APIs (#633)
palantir/godel-conjure-plugin: README.md

* `conjure`: runs Conjure generation. Runs for all of the entries specified in the configuration in order. The working
  directory is set to be the project directory.
* `conjure-publish`: publishes IR to a specified destination.
* `conjure-backcompat`: checks backward compatibility of Conjure APIs by comparing the current IR against the previously
  published version. Runs automatically as part of `./godelw verify`. Projects without a `group-id` or with `publish: false`

>   are skipped.

Reviewer: Let's go with `check-conjure-backcompat` for consistency with other tooling

> * `conjure-accept-backcompat-breaks`: accepts backward compatibility breaks by writing acknowledgment entries to lockfiles.

Reviewer: ```suggestion
* `conjure-accept-backcompat-breaks`: accepts any compatibility breaks identified by `conjure-backcompat` (if supported by the asset).
```

Similarly to the above, the generic contract at the plugin level is "accepts any compatibility breaks identified by `conjure-backcompat` (if supported by the asset)."

Probably fine to include this, although now that this is written out another possibility would be for an asset to specify whether or not it supports the concept of "accepting" backcompat breaks.

> * `conjure-accept-backcompat-breaks`: accepts backward compatibility breaks by writing acknowledgment entries to lockfiles.

Reviewer: Let's go with `accept-conjure-breaks` for consistency with other tooling

- The asset must respond to the `_assetInfo` probe with:
  ```json
  { "type": "backcompat" }
  ```

**Invocation:**
- The asset is invoked with a JSON object that uses a discriminated union pattern based on operation type.

**Check Backcompat Mode (`checkBackCompat`):**
```json
{
  "type": "checkBackCompat",
  "checkBackCompat": {
    "currentIR": "/tmp/current-ir.json",
    "project": "my-api",
    "groupId": "com.example",
    "projectConfig": {
      "output-dir": "./pkg",
      "server": true,
      "ir-locator": { "type": "yaml", "locator": "./conjure" }
    },
    "godelProjectDir": "/path/to/project"
  }
}
```

**Accept Backcompat Breaks Mode (`acceptBackCompatBreaks`):**
```json
{
  "type": "acceptBackCompatBreaks",
  "acceptBackCompatBreaks": {
    "currentIR": "/tmp/current-ir.json",
    "project": "my-api",
    "groupId": "com.example",
    "projectConfig": {
      "output-dir": "./pkg",
      "server": true,
      "ir-locator": { "type": "yaml", "locator": "./conjure" }
    },
    "godelProjectDir": "/path/to/project"
  }
}
```

### Usage

**Running Backcompat Checks:**
```bash
# Check all projects
./godelw conjure-backcompat

# Check a specific project

> ./godelw conjure-backcompat --project=my-api

Reviewer: I agree that this is behavior that we want to support, but it looks like none of the other commands in the plugin currently support specifying projects.

For this initial PR, let's match the other behavior of just doing it for all projects.

In a follow-up PR, we can add per-project support for all commands.

Also, for most user-facing godel commands, targets tend to be arguments rather than flags, so would likely prefer to go with that approach for consistency.

```

**Accepting Backcompat Breaks:**
```bash
# Accept breaks for all projects
./godelw conjure-accept-backcompat-breaks

# Accept breaks for a specific project
./godelw conjure-accept-backcompat-breaks --project=my-api
```

### Behavior

- **Non-published projects**: Projects without a `group-id` configured or with `publish: false` are skipped.
- **No backcompat asset**: If no backcompat asset is configured, the tasks are no-ops.
- **Multiple assets**: Only one backcompat asset is supported. If multiple backcompat assets are configured, the plugin will fail with an error.

### Error Handling

- If the asset is invoked with more than one argument, it must immediately fail.
- On success, the asset should output `{}` and exit with code 0.

> - On failure, the asset should output an error message to stdout and exit with a non-zero code.

Reviewer: Behavior of "output empty JSON on success, but output non-JSON error message on failure" seems counter-intuitive as it is not symmetric.

Would expect either:
1. Output nothing on success with exit code 0, error message on failure with non-0 exit code (preferred)
2. Always output JSON, in which case success can either be empty or have a field like "success/fail" and failure would set a field like "errorMessage").
